### ISSUE TITLE
Slow SSL Authentication on Ubuntu 16.04 / 18.04

### ISSUE THREAD

**Issue by karmeli87** [dotnet/runtime#27666]
We investigate errors during SSL authentication on ubuntu using kestrel. 
We initiate many concurrent connections and observer that we hang on `AuthenticateAsClientAsync` and `AuthenticateAsServerAsync`.
While trying to make a minimal repo, we decided to do a simple sever-client app that simulate our usage.

Running the same app on windows resulted in:
Plain ~20ms
SSL ~200ms

while on ubuntu 18.04:
Plain ~1s
SSL ~40s 

I also tried on a VM running 16.04
Plain ~1s
SSL ~5s

In all of the test the dotnet version is `2.1.403`
```c#
using System;
using System.Net;
using System.Diagnostics;
using System.Net.Sockets;
using System.Security.Cryptography.X509Certificates;
using System.Threading;
using System.Threading.Tasks;

namespace ConsoleApp1
{
    public static class Program
    {
        public static void Main(string[] args)
        {
#if USE_SSL
        System.Console.WriteLine("SSL");
#else
            Console.WriteLine("Plain");
#endif
            var globalCertificate = new X509Certificate2("path/to/cert.pfx");
            var listener = new TcpListener(new IPAddress(new byte[] { 127, 0, 0, 1 }), 56789);
            listener.Start();
            for (var i = 0; i < 1; i++)
                Listen(listener, globalCertificate);
            var overall = Stopwatch.StartNew();
            int done = 0;
            int failed = 0;
            var clients = 200;
            var tasks = new Task[clients];
            for (int i = 0; i < clients; i++)
            {
                var j = i;
                tasks[i] = Task.Run(async () =>
                {
                    var sp = Stopwatch.StartNew();
                    try
                    {
                        using (var tcpClient = new TcpClient())
                        {
                            await tcpClient.ConnectAsync(new IPAddress(new byte[] { 127, 0, 0, 1 }), 56789);
#if USE_SSL
                            using (var wrapped = new SslStream(tcpClient.GetStream(), false, ((sender, certificate, chain, errors) => true)))
                            {
                                await wrapped.AuthenticateAsClientAsync("localhost", new X509CertificateCollection(new[] {globalCertificate}),
                                    SslProtocols.Tls12, false);
#else
                            using (var wrapped = tcpClient.GetStream())
                            {
#endif
                                byte[] buffer = { 1 };
                                await wrapped.WriteAsync(buffer, 0, 1);
                                var read = await wrapped.ReadAsync(buffer, 0, 1);

                                if (read != 1)
                                {
                                    Console.WriteLine("why?");
                                }

                                Interlocked.Increment(ref done);
                            }
                        }
                    }
                    catch (Exception e)
                    {
                        Console.WriteLine(Interlocked.Increment(ref failed) + " FAILED " + sp.Elapsed);
                        Console.WriteLine(e);
                    }
                });
            }

            Task.WaitAll(tasks);

            Console.WriteLine("Done " + done + " failed " + failed + " in " + overall.Elapsed);
        }

        private static void Listen(TcpListener listener, X509Certificate globalCertificate)
        {
            Task.Run(async () =>
            {

                try
                {
                    var tcpClient = await listener.AcceptTcpClientAsync();
                    Listen(listener, globalCertificate);

#if USE_SSL
                    using (var wrapped = new SslStream(tcpClient.GetStream(), false, ((sender, certificate, chain, errors) => true)))
                    {
                        await wrapped.AuthenticateAsServerAsync(globalCertificate);
#else
                    using (var wrapped = tcpClient.GetStream())
                    {
#endif

                        var buffer = new byte[1];

                        var read = await wrapped.ReadAsync(buffer, 0, 1);

                        await wrapped.WriteAsync(buffer, 0, 1);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e);
                }
            });
        }
    }
}
```

**Comment by wfurt** [comment]
This should be fixed in 7.0
running benchmark from https://github.com/dotnet/performance

```
Ubuntu 20.04
|                           Method |        Job |  Runtime | Toolchain | protocol |         Mean |        Error |       StdDev |       Median |           Min |          Max | Ratio | RatioSD | Allocated | Alloc Ratio |
|--------------------------------- |----------- |--------- |---------- |--------- |-------------:|-------------:|-------------:|-------------:|--------------:|-------------:|------:|--------:|----------:|------------:|
| DefaultHandshakeContextIPv4Async | Job-TGUDPL | .NET 6.0 |    net6.0 |        ? | 7,140.94 us |  72.833 us |  68.128 us | 7,122.85 us | 7,049.24 us | 7,281.16 us |  1.00 |    0.00 |   19856 B |        1.00 |
| DefaultHandshakeContextIPv4Async | Job-GVAJIU | .NET 7.0 |    net7.0 |        ? | 2,226.72 us |  44.359 us |  39.323 us | 2,227.98 us | 2,141.81 us | 2,297.37 us |  0.31 |    0.01 |    8835 B |        0.44 |

Windows Server 2022
|                           Method |        Job |  Runtime | Toolchain | protocol |         Mean |        Error |       StdDev |       Median |           Min |          Max | Ratio | RatioSD | Allocated | Alloc Ratio |
|--------------------------------- |----------- |--------- |---------- |--------- |-------------:|-------------:|-------------:|-------------:|--------------:|-------------:|------:|--------:|----------:|------------:|
| DefaultHandshakeContextIPv4Async | Job-FEYPRU | .NET 6.0 |    net6.0 |        ? |  2,409.34 us |  16.292 us |  15.239 us |  2,407.36 us |  2,388.792 us |  2,438.48 us |  1.00 |    0.00 |   13915 B |        1.00 |
| DefaultHandshakeContextIPv4Async | Job-VJNBCK | .NET 7.0 |    net7.0 |        ? |  2,467.38 us |  71.931 us |  79.952 us |  2,421.02 us |  2,391.125 us |  2,650.48 us |  1.03 |    0.04 |    4811 B |        0.35 |
```

For TLS resume to work, the server side has to use  SslStreamCertificateContext. (and no CipherSuitePolicy)

**Comment by wfurt** [comment]
one more benchmark -> `crank` form asp.net
RPS:  1,018  ->19,928


crank compare 6.0.json 7.0.json
 
| application           | 6.0     | 7.0                   |         |
| --------------------- | ------- | --------------------- | ------- |
| CPU Usage (%)         |      70 |                    78 | +11.43% |
| Cores usage (%)       |     845 |                   930 | +10.06% |
| Working Set (MB)      |     182 |                   286 | +57.14% |
| Private Memory (MB)   |     679 |                   862 | +26.95% |
| Build Time (ms)       |   3,208 |                 1,388 | -56.73% |
| Start Time (ms)       |     149 |                   232 | +55.70% |
| Published Size (KB)   | 112,861 |               113,797 |  +0.83% |
| .NET Core SDK Version | 6.0.400 | 7.0.100-rc.1.22423.18 |         |
 
 
| load                   | 6.0    | 7.0     |            |
| ---------------------- | ------ | ------- | ---------- |
| CPU Usage (%)          |    100 |      99 |     -1.00% |
| Cores usage (%)        |  1,201 |   1,187 |     -1.17% |
| Working Set (MB)       |    109 |      49 |    -55.05% |
| Private Memory (MB)    |    460 |     357 |    -22.39% |
| Start Time (ms)        |      0 |       0 |            |
| First Request (ms)     |    199 |     212 |     +6.53% |
| Requests/sec           |  1,018 |  19,928 | +1,857.25% |
| Requests               | 15,365 | 300,772 | +1,857.51% |
| Mean latency (ms)      |   2.81 |    0.56 |    -80.05% |
| Max latency (ms)       |  37.77 |   31.87 |    -15.62% |
| Bad responses          |      0 |       0 |            |
| Socket errors          |      0 |       0 |            |
| Read throughput (MB/s) |   0.15 |    2.87 | +1,857.43% |
| Latency 50th (ms)      |   0.73 |    0.30 |    -59.28% |
| Latency 75th (ms)      |   4.24 |    0.51 |    -88.07% |
| Latency 90th (ms)      |   9.92 |    0.77 |    -92.26% |
| Latency 99th (ms)      |  17.78 |    8.00 |    -55.01% |
